src/app/qgsdockablewidget.cpp

```diff
@@ -0,0 +1,119 @@
+/***************************************************************************
+  qgsdockablewidget.cpp
+  --------------------------------------
+  Date                 : July 2017
+  Copyright            : (C) 2017 by Martin Dobias
+  Email                : wonder dot sk at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsdockablewidget.h"
+
+#include "qgisapp.h"
+#include "qgs3dmapcanvaswidget.h"
+#include "qgsdockwidget.h"
+
+#include <QWidget>
+
+QgsDockableWidget::QgsDockableWidget( QWidget *parent )
+  : QWidget( parent )
+{
+  this->setAttribute( Qt::WA_DeleteOnClose );
+
+  mDock = new QgsDockWidget( this );
+  mDock->setAttribute( Qt::WA_DeleteOnClose );
+
+  mDock->setWidget( nullptr );
+  mDock->setAllowedAreas( Qt::AllDockWidgetAreas );
+
+  mDock->setVisible( false );
+
+  mDialog = new QDialog( this, Qt::Window );
```

Comment:
I don't understand why we are parenting a top-level QDialog to QgsDockableWidget (and the same for QgsDockWidget)...

Do we even need QgsDockableWidget to be a QWidget? I am a bit lost where it belongs in the hierarchy and if it actually gets shown. Maybe a simpler design would be to make this a kind of utility class (e.g. `QgsDockableWidgetHelper`) and the 3D map canvas would have it as a member variable? That could also simplify the code in qgisapp.cpp where the functions return a generic `QgsDockableWidget` which is a bit inconvenient...

Reply:
Actually I wanted to make QgsDockableWidget a QWidget in order to be able to hide/show both the dialog and the dock widget and also handle closing and deletion of both at the same time the QgsDockableWidget is closed or deleted.
One issue that I found was that the QgsDockWidget was always parented to the QgisApp::instance() which makes things more inconvenient.
I thought about making the allocation and connections linking outside of the QgsDockableWidget but it appears to me as the more complicated approach and we ultimately want a black box widget that we give a widget and it takes care of putting it wherever we tell it to.